relation_engine/taxa/rdp/parsers.py

```diff
@@ -0,0 +1,159 @@
+"""
+Common code for dealing with RDP taxonomy files.
+"""
+
+# TODO TEST
+# TODO DOCS better documentation.
+
+import re
+
+_16S = '16S'
+_28S = '28S'
+_INCERTAE_SEDIS = 'incertae_sedis'
+
+_RE_INCERTAE_SEDIS = re.compile('[_ ][Ii]ncertae[_ ][Ss]edis')
+
+class RDPNodeProvider:
+    """
+    RDPNodeProvider is an iterable that returns a new RDP taxonomy node as a dict with each
+    iteration.
+    """
+
+    def __init__(self, rdp_taxonomy_16Sfile_handles, rdp_taxonomy_28Sfile_handles):
+        """
+        Create the node provider.
+
+        rdp_taxonomy_16Sfile_handles - a list of open handles for the RDP taxonomy 16S files to
+            process.
+        rdp_taxonomy_28Sfile_handles - a list of open handles for the RDP taxonomy 28S files to
+            process.
+        """
+        self._fh_16S = rdp_taxonomy_16Sfile_handles
+        self._fh_28S = rdp_taxonomy_28Sfile_handles
+
+    def __iter__(self):
+        seen_taxa = set()  # not including leaves
+        for fh in self._fh_16S:
+            yield from self._processfile(_16S, fh, seen_taxa)
+        for fh in self._fh_28S:
+            yield from self._processfile(_28S, fh, seen_taxa)
+        
+    def _processfile(self, molecule, fh, seen_taxa):
+        for line in fh:
+            if not line.startswith('>'):
+                continue
+            names, lineage = line.split('\t')
+            locus, definition = names.split(' ', 1)
+            lineage, unclassified = _get_lineage(lineage)
+            if not lineage:  # it's an outgroup
+                continue
+            for l in lineage:
+                l_id = _taxon_to_id(l)
+                if l_id not in seen_taxa:
+                    yield {
+                        'id': l_id.replace('/', '_'),
+                        'rank': l['rank'],
+                        'name': l['name'],
+                        'unclassified': False,
+                        'molecule': None,
+                        _INCERTAE_SEDIS: l[_INCERTAE_SEDIS]
+                    }
+                seen_taxa.add(l_id)
+            yield {
+                'id': locus[1:].strip(),  # remove '>'
+                'rank': 'sequence_example',
+                'name': definition.strip(),
+                'unclassified': unclassified,
+                'molecule': molecule,
+                _INCERTAE_SEDIS: None
+            }
+
+class RDPEdgeProvider:
```

Comment:
Why not make this a single generator function?

Reply:
Because the algorithm requires all the nodes are loaded before the edges, so I need a separate iterators for each